pkg/controller/quota/quota.go

```diff
@@ -0,0 +1,165 @@
+package quota
+
+import (
+	"fmt"
+
+	v1 "github.com/acorn-io/acorn/pkg/apis/api.acorn.io/v1"
+	apiv1 "github.com/acorn-io/acorn/pkg/apis/internal.acorn.io/v1"
+	adminv1 "github.com/acorn-io/acorn/pkg/apis/internal.admin.acorn.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/acorn-io/acorn/pkg/condition"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"github.com/acorn-io/baaah/pkg/router"
+)
+
+// WaitForAllocation blocks the appInstance from being deployed until quota has been allocated on
+// an associated QuotaRequest object.
+func WaitForAllocation(req router.Request, resp router.Response) error {
+	appInstance := req.Object.(*apiv1.AppInstance)
+
+	// Create a condition setter for AppInstanceConditionQuotaAllocated, which blocks the appInstance from being deployed
+	// until quota has been allocated.
+	status := condition.Setter(appInstance, resp, apiv1.AppInstanceConditionQuotaAllocated)
+
+	// Don't do anything if quota isn't enabled for this project
+	if enforced, err := isEnforced(req, appInstance.Namespace); !enforced {
+		if err != nil {
+			status.Error(err)
+			return err
+		}
+		status.Success()
+		return nil
+	}
+
+	// Attempt to get the quotaRequest for this appInstance. It should exist with the name and namespace of the
+	// appInstance being processed.
+	quotaRequest := &adminv1.QuotaRequestInstance{}
+	err := req.Client.Get(req.Ctx, router.Key(appInstance.Namespace, appInstance.Name), quotaRequest)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// Determine how to proceed depending on if the quotaRequest exists and what it has written to its status
+	if quotaRequest.Status.FailedResources != nil {
```

Comment:
This part of the code takes me a while to understand.. would be nice to add some comments explaining each scenario? It is always good if you have a big if else statement and comment helps a lot in understanding it.

Reply:
Great point. Let me add that and demystify this code a bit.

Reply:
💯